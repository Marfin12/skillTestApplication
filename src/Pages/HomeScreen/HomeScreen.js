import * as React from 'react';
import { View, Image } from 'react-native';
import { graphql, compose } from 'react-apollo';

import { getSportList } from '../../Graphql/sport.graphql';

import Card from '../../Components/Card';
import ListView from '../../Components/ListView';
import HeaderList from '../../Components/HeaderList';
import TextField from '../../Components/TextField';

import styles from './HomeScreen.styles';
import { navigateToDetailScreen } from './HomeScreen.utils';

const _renderHeader = () => <HeaderList>List View Header</HeaderList>;

const _renderItemContent = (name, description) => (
  <View style={styles.containerText}>
    <TextField textStyle={styles.title}> {name} </TextField>
    <TextField numberOfLines={3} style={styles.description}>
    {' '}
      {description}{' '}
    </TextField>
  </View>
);

export const _renderItem = (props, navigation) => {
  const { name, description, image } = props;

  return (
    <Card
      onPress={navigateToDetailScreen(navigation, props)}>
        <Image source={{uri: image}} style={styles.photo} />
        {_renderItemContent(name, description)}
    </Card>
  );
};

const _renderEmpty = () => (
  <TextField style={styles.emptyList} textStyle={styles.description}>
    The list data is empty
  </TextField>
);

const HomeScreen = props => {
  return (
  <React.Fragment>
    <ListView
      style={styles.listView}
      item={props.data.sports}
      navigation={props.navigation}
      itemList={_renderItem}
      emptyList={_renderEmpty}
      headerList={_renderHeader}
    />
  </React.Fragment>
);
  };

export default compose(graphql(getSportList))(HomeScreen);
