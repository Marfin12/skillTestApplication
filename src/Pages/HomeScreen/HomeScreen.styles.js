// @flow

const styles = {
  title: {
    fontSize: 20,
  },
  containerText: {
    flex: 1,
    flexDirection: 'column',
    marginLeft: 12,
    marginTop: -10,
    justifyContent: 'center',
  },
  description: {
    fontSize: 11,
    fontStyle: 'italic',
    marginTop: 10,
  },
  photo: {
    height: 80,
    width: 80,
  },
  emptyList: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  listView: {
    marginTop: 20,
  },
};

export default styles;
