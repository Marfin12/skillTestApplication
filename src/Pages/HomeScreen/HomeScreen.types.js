// @flow

import type {
  ViewStyleProp
} from 'react-native/Libraries/StyleSheet/StyleSheet';

import Constants from '../../Constants';

export type Item = {|
  name: string,
  description: string,
  image: string
|};

export type Data = {|
  sports: Item
|};

export type Props = {|
  data: Data,
  navigation: () => void
|};
