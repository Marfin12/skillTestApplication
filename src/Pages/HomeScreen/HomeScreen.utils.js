const navigateToDetailScreen = (navigation, props) => () => {
  console.log(navigation);
  navigation.navigate('Details', {props});
};

const apolloObjectCreator = sports => ({
  length: sports.length,
  sports: sports,
});

const apolloMapCreator = () => {
  const queryStore = new Map();
  queryStore.set({
    queryId: 123,
  });

  return queryStore;
};

export {navigateToDetailScreen, apolloObjectCreator, apolloMapCreator};
