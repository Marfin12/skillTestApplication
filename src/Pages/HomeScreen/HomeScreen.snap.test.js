import React from 'react';
import renderer from 'react-test-renderer';
import { ApolloProvider } from 'react-apollo';

import HomeScreen from './HomeScreen';
import { apolloObjectCreator, apolloMapCreator } from './HomeScreen.utils';

jest.mock('../../Components/TextField', () => 'TextField');
jest.mock('../../Components/Card', () => 'Card');
jest.mock('../../Components/HeaderList', () => 'Card');

const navigation = jest.fn();
const client = {
  watchQuery: jest.fn(),
  refetch: false,
};
const mockedApolloConfig = {
  refetch: jest.fn(),
  fetchMore: jest.fn(),
  updateQuery: jest.fn(),
  startPolling: jest.fn(),
  stopPolling: jest.fn(),
  subscribeToMore: jest.fn(),
  options: {
    fetchPolicy: 'no-cache',
  },
  queryManager: {
    queryStore: apolloMapCreator(),
  },
  queryId: 123,
  subscribe: jest.fn(),
  currentResult: {
    loading: false,
    partial: true,
    networkStatus: 'ready',
    errors: null,
    error: null,
  },
};

let HomeComponent;

afterEach(() => {
  jest.clearAllMocks();
});

describe('Home Screen Component with loading', () => {
  jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    const lightMode = {
      theme: false,
    };
    const darkMode = {
      theme: true,
    };
    return {
      ...ActualReactRedux,
      useSelector: jest
        .fn()
        .mockImplementationOnce(() => {
          return lightMode;
        })
        .mockImplementationOnce(() => {
          return darkMode;
        }),
    };
  });

  beforeEach(() => {
    client.watchQuery = jest.fn(() => ({
      ...mockedApolloConfig,
      currentResult: jest.fn().mockImplementation(() => ({
        ...mockedApolloConfig.currentResult,
        loading: true,
      })),
    }));
  });

  HomeComponent = (
    <ApolloProvider client={client}>
      <HomeScreen navigation={navigation} />
    </ApolloProvider>
  );

  test('renders light mode home screen', () => {
    const ScreenWrapper = renderer.create(HomeComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });

  test('renders dark mode home screen', () => {
    const ScreenWrapper = renderer.create(HomeComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });
});

describe('Home Screen Component with failed to fetch', () => {
  jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    const lightMode = {
      theme: false,
    };
    const darkMode = {
      theme: true,
    };
    return {
      ...ActualReactRedux,
      useSelector: jest
        .fn()
        .mockImplementationOnce(() => {
          return lightMode;
        })
        .mockImplementationOnce(() => {
          return darkMode;
        }),
    };
  });

  beforeEach(() => {
    const sports = [];

    client.watchQuery = jest.fn(() => ({
      variables: sports,
      ...mockedApolloConfig,
      currentResult: jest.fn().mockImplementation(() => ({
        ...mockedApolloConfig.currentResult,
        data: apolloObjectCreator(sports),
      })),
    }));
  });

  HomeComponent = (
    <ApolloProvider client={client}>
      <HomeScreen navigation={navigation} />
    </ApolloProvider>
  );

  test('renders light mode home screen', () => {
    const ScreenWrapper = renderer.create(HomeComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });

  test('renders dark mode home screen', () => {
    const ScreenWrapper = renderer.create(HomeComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });
});

describe('Home Screen Component with successfully fetch', () => {
  jest.mock('react-redux', () => {
    const ActualReactRedux = jest.requireActual('react-redux');
    const lightMode = {
      theme: false,
    };
    const darkMode = {
      theme: true,
    };
    return {
      ...ActualReactRedux,
      useSelector: jest
        .fn()
        .mockImplementationOnce(() => {
          return lightMode;
        })
        .mockImplementationOnce(() => {
          return darkMode;
        }),
    };
  });

  beforeEach(() => {
    const sports = [
      {
        name: 'Sport Example',
        image: 'Some Image',
        description: 'This is sport description...',
      },
    ];

    client.watchQuery = jest.fn(() => ({
      variables: sports,
      ...mockedApolloConfig,
      currentResult: jest.fn().mockImplementation(() => ({
        ...mockedApolloConfig.currentResult,
        data: apolloObjectCreator(sports),
      })),
    }));
  });

  HomeComponent = (
    <ApolloProvider client={client}>
      <HomeScreen navigation={navigation} />
    </ApolloProvider>
  );

  test('renders light mode home screen', () => {
    const ScreenWrapper = renderer.create(HomeComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });

  test('renders dark mode home screen', () => {
    const ScreenWrapper = renderer.create(HomeComponent).toJSON();
    expect(ScreenWrapper).toMatchSnapshot();
  });
});
