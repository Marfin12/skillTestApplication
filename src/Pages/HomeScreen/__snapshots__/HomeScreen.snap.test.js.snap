// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen Component with failed to fetch renders dark mode home screen 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 16,
      },
      Object {
        "marginTop": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <TextField
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    }
    ListHeaderComponent={
      <Card>
        List View Header
      </Card>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Card>
          List View Header
        </Card>
      </View>
      <TextField
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Home Screen Component with failed to fetch renders light mode home screen 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 16,
      },
      Object {
        "marginTop": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <TextField
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    }
    ListHeaderComponent={
      <Card>
        List View Header
      </Card>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Card>
          List View Header
        </Card>
      </View>
      <TextField
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Home Screen Component with loading renders dark mode home screen 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 16,
      },
      Object {
        "marginTop": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <TextField
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    }
    ListHeaderComponent={
      <Card>
        List View Header
      </Card>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Card>
          List View Header
        </Card>
      </View>
      <TextField
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Home Screen Component with loading renders light mode home screen 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 16,
      },
      Object {
        "marginTop": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <TextField
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    }
    ListHeaderComponent={
      <Card>
        List View Header
      </Card>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Card>
          List View Header
        </Card>
      </View>
      <TextField
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Home Screen Component with successfully fetch renders dark mode home screen 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 16,
      },
      Object {
        "marginTop": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <TextField
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    }
    ListHeaderComponent={
      <Card>
        List View Header
      </Card>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={
      Array [
        Object {
          "description": "This is sport description...",
          "image": "Some Image",
          "name": "Sport Example",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Card>
          List View Header
        </Card>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Card
          onPress={[Function]}
        >
          <Image
            source={
              Object {
                "uri": "Some Image",
              }
            }
            style={
              Object {
                "height": 80,
                "width": 80,
              }
            }
          />
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
                "marginLeft": 12,
                "marginTop": -10,
              }
            }
          >
            <TextField
              textStyle={
                Object {
                  "fontSize": 20,
                }
              }
            >
               
              Sport Example
               
            </TextField>
            <TextField
              numberOfLines={3}
              style={
                Object {
                  "fontSize": 11,
                  "fontStyle": "italic",
                  "marginTop": 10,
                }
              }
            >
               
              This is sport description...
               
            </TextField>
          </View>
        </Card>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Home Screen Component with successfully fetch renders light mode home screen 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 16,
      },
      Object {
        "marginTop": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <TextField
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
        textStyle={
          Object {
            "fontSize": 11,
            "fontStyle": "italic",
            "marginTop": 10,
          }
        }
      >
        The list data is empty
      </TextField>
    }
    ListHeaderComponent={
      <Card>
        List View Header
      </Card>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={
      Array [
        Object {
          "description": "This is sport description...",
          "image": "Some Image",
          "name": "Sport Example",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <Card>
          List View Header
        </Card>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Card
          onPress={[Function]}
        >
          <Image
            source={
              Object {
                "uri": "Some Image",
              }
            }
            style={
              Object {
                "height": 80,
                "width": 80,
              }
            }
          />
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
                "marginLeft": 12,
                "marginTop": -10,
              }
            }
          >
            <TextField
              textStyle={
                Object {
                  "fontSize": 20,
                }
              }
            >
               
              Sport Example
               
            </TextField>
            <TextField
              numberOfLines={3}
              style={
                Object {
                  "fontSize": 11,
                  "fontStyle": "italic",
                  "marginTop": 10,
                }
              }
            >
               
              This is sport description...
               
            </TextField>
          </View>
        </Card>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
